contracts/oracles/yieldspace/PoolOracle.sol

```diff
@@ -0,0 +1,146 @@
+                // no overflow issue. if observationIndex + 1 overflows, result is still zero.
+                oldestObservationIndex = (++observationIndex) % granularity;
+                // If no data exists for this index, check the next one,
+                // this should only happen during the first timeWindow for a given pool
+                // If the elapsedTime is bigger than the timeWindow,
+                // we also check the next one in case only this particular observation is stale
+                if (block.timestamp - poolObservations[pool][oldestObservationIndex].timestamp < windowSize) {
+                    return poolObservations[pool][oldestObservationIndex];
+                }
+            }
+
+            revert MissingHistoricalObservation(pool);
+        }
+    }
+
+    // @inheritdoc IPoolOracle
+    function update(address pool) public override {
+        // populate the array with empty observations (oldest call only)
+        for (uint256 i = poolObservations[pool].length; i < granularity; i++) {
+            poolObservations[pool].push();
+        }
+
+        // get the observation for the current period
+        uint256 index = observationIndexOf(block.timestamp);
+        Observation storage observation = poolObservations[pool][index];
+
+        // we only want to commit updates once per period (i.e. windowSize / granularity)
+        uint256 timeElapsed = block.timestamp - observation.timestamp;
+        if (timeElapsed > periodSize) {
+            observation.timestamp = block.timestamp;
+            observation.ratioCumulative = _getCurrentCumulativeRatio(pool);
+            emit ObservationRecorded(pool, index, observation);
+        }
+    }
+
+    /// @inheritdoc IPoolOracle
+    function peek(address pool) public view override returns (uint256 twar) {
+        Observation memory oldestObservation = getOldestObservationInWindow(pool);
+
+        uint256 timeElapsed = block.timestamp - oldestObservation.timestamp;
+        if (timeElapsed > windowSize) {
+            revert MissingHistoricalObservation(pool);
+        }
+
+        if (timeElapsed < minTimeElapsed) {
+            revert InsufficientElapsedTime(pool, timeElapsed);
+        }
+
+        // cumulative ratio is in (ratio * seconds) units so for the average we simply get it after division by time elapsed
+        return ((_getCurrentCumulativeRatio(pool) - oldestObservation.ratioCumulative) * 1e18) / (timeElapsed * 1e27);
+    }
+
+    /// @inheritdoc IPoolOracle
+    function get(address pool) external override returns (uint256 twar) {
+        update(pool);
+        return peek(pool);
+    }
+
+    /// @dev computes the most up-to-date TWAR based last calculated ratio and the current cached balances on the pool
+    /// @param pool Address of pool for which the ratio is required 
+    /// @return lastRatio The latest TWAR for `pool`
+    function _getCurrentCumulativeRatio(address pool) internal view returns (uint256 lastRatio) {
```

Comment:
can we consider a different name, like `currentRatio` -- to me this sounds too close to `cumulativeRatioLast` which was the initial source of confusion.  i'm so scared of that thing now after the bad experience we had the first time we made this oracle, i even put this warning in the new Pool.sol contract lol:
![Screen Shot 2022-05-25 at 3 05 21 PM](https://user-images.githubusercontent.com/71567643/170375053-9d3d1be2-bdf6-420e-8f00-7ef9d7bbaa1e.png)

Reply:
btw as you can see we are providing a new thing that you can use `currentCumulativeRatio` in the new pool contract.  there are a couple of breaking changes as well.  and also we are planning to deploy the new contracts in June.  but i want to get through this first set of pr comments on the oracle here before we look at those changes related to the new pool contract.

Reply:
Yes, this code should be in the pool itself imho, or ate the very least in a provided lib like the one you posted in another comment. That way people wouldn't have the same issues we did.

Reply:
yup